workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,214 @@
+                {
+                    var id = idToReaderWriterList.Key;
+                    var readerWriterList = idToReaderWriterList.Value;
+                    foreach (var readerWriter in readerWriterList)
+                    {
+                        compIdToReadersWriters[id].Remove(readerWriter);
+                    }
+                }
+
+                behaviourToReadersWriters.Remove(behaviour);
+            }
+
+            behavioursToDisable.Clear();
         }
 
         public void AddComponent(uint componentId)
         {
-            // Mark readers as to be enabled
+            // Mark reader components ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentRequirementSatisfied(relevantReaderSpatialOSBehaviours);
         }
 
         public void RemoveComponent(uint componentId)
         {
-            // Mark readers as to be disabled
+            // Mark reader components not ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentRequirementUnsatisfied(relevantReaderSpatialOSBehaviours);
         }
 
         public void ChangeAuthority(uint componentId, Authority authority)
         {
-            // Mark writers as to be enabled/disabled
+            if (authority == Authority.Authoritative)
+            {
+                // Mark writer components ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = behavioursRequiringWriterTypes[componentId];
+                MarkComponentRequirementSatisfied(relevantWriterSpatialOSBehaviours);
+            }
+            else if (authority == Authority.NotAuthoritative)
+            {
+                // Mark writer components not ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = behavioursRequiringWriterTypes[componentId];
+                MarkComponentRequirementUnsatisfied(relevantWriterSpatialOSBehaviours);
+            }
+        }
+
+        private void MarkComponentRequirementSatisfied(IEnumerable<MonoBehaviour> behaviours)
+        {
+            // Inject all Readers/Writers at once when all requirements are met
+            foreach (var behaviour in behaviours)
+            {
+                numUnsatisfiedReadersOrWriters[behaviour]--;
+                if (numUnsatisfiedReadersOrWriters[behaviour] == 0)
+                {
+                    if (!behavioursToEnable.Contains(behaviour))
+                    {
+                        // Schedule activation
+                        behavioursToEnable.Add(behaviour);
+                    }
+                    else
+                    {
+                        // Must be enabled already, so we were going to disable it - let's not
+                        behavioursToDisable.Remove(behaviour);
```

Comment:
Are you sure the code here is correct? You are saying "If the behaviour was previously already scheduled to be enabled, we need to remove it from the list of behaviours to be disabled". Shouldn't this rather be:
```
if (numUnsatisfiedReadersOrWriters[behaviour] == 0)
{
    if (!behavioursToEnable.Contains(behaviour))
    {
        behavioursToEnable.Add(behaviour);
    }
    else
    {
        LogError("THis shouldn't happen");
    }

    if (!behavioursToDisable.Contains(behaviour))
    {
        behavioursToDisable.Remove(behaviour);
    }
}
```